smart_open/asb.py

```diff
@@ -0,0 +1,410 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2020 Nicolas Mitchell <ncls.mitchell@gmail.com>
+#
+# This code is distributed under the terms and conditions
+# from the MIT License (MIT).
+#
+"""Implements file-like objects for reading and writing to/from Azure Storage Blob (ASB)."""
+
+import io
+import logging
+import sys
+
+from azure.storage.blob import BlobServiceClient, StorageStreamDownloader
+from azure.core.exceptions import ResourceNotFoundError
+import six
+
+import smart_open.bytebuffer
+
+logger = logging.getLogger(__name__)
+
+_READ_BINARY = 'rb'
+_WRITE_BINARY = 'wb'
+
+_BINARY_TYPES = (six.binary_type, bytearray)
+"""Allowed binary buffer types for writing to the underlying Azure Storage Blob stream"""
+
+SUPPORTED_SCHEME = "asb"
+"""Supported scheme for Azure Storage Blob in smart_open endpoint URL"""
+
+_MIN_MIN_PART_SIZE = _REQUIRED_CHUNK_MULTIPLE = 4 * 1024**2
+"""Azure requires you to upload in multiples of 4MB, except for the last part."""
+
+_DEFAULT_MIN_PART_SIZE = 64 * 1024**2
+"""Default minimum part size for Azure Cloud Storage multipart uploads is 64MB"""
+
+DEFAULT_BUFFER_SIZE = 4 * 1024**2
+"""Default buffer size for working with Azure Storage Blob is 256MB
+https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs
+"""
+
+START = 0
+"""Seek to the absolute start of an Azure Storage Blob file"""
+
+CURRENT = 1
+"""Seek relative to the current positive of an Azure Storage Blob file"""
+
+END = 2
+"""Seek relative to the end of an Azure Storage Blob file"""
+
+_WHENCE_CHOICES = (START, CURRENT, END)
+
+_BINARY_NEWLINE = b'\n'
+
+def open(
+        bucket_id,
+        blob_id,
+        mode,
+        buffer_size=DEFAULT_BUFFER_SIZE,
+        client=None,  # type: azure.storage.blob.BlobServiceClient
+        ):
+    """Open an Azure Storage Blob blob for reading or writing.
+
+    Parameters
+    ----------
+    bucket_id: str
+        The name of the bucket this object resides in.
+    blob_id: str
+        The name of the blob within the bucket.
+    mode: str
+        The mode for opening the object.  Must be either "rb" or "wb".
+    buffer_size: int, optional
+        The buffer size to use when performing I/O. For reading only.
+    client: azure.storage.blob.BlobServiceClient, optional
+        The Azure Storage Blob client to use when working with azure-storage-blob.
+
+    """
+    if mode == _READ_BINARY:
+        return SeekableBufferedInputBase(
```

Comment:
call this class Reader, and the other one Writer. Please see the master branch (s3 and gcs implementations)